File: src/single_world/python/abstraction/mpmd_manager.py
from mpi4py import MPI

from .program_data import ProgramData
from .filename_handler import get_filename

class MPMDManager:
  __local_data: ProgramData
  __manager_comm = MPI.COMM_NULL
  __programs_data: list[ProgramData] = []

  @staticmethod
  def initialize():
    MPMDManager.__manager_comm = MPI.COMM_WORLD.Dup()

    local_name = get_filename()
    gathered_names = MPMDManager.__manager_comm.allgather(
      local_name
    )
    MPMDManager.__set_programs_data(local_name, gathered_names)

  @staticmethod
  def local_name(): return MPMDManager.__local_data.name

  @staticmethod
  def local_comm() -> MPI.Intracomm: return MPMDManager.__local_data.comm
  
  @staticmethod
  def local_size(): return MPMDManager.__local_data.comm.Get_size()

  @staticmethod
  def intercomm_to(identifier: str | int):
    return MPMDManager.__find_program_or_error(identifier).comm

  @staticmethod
  def size_of(identifier: str | int):
    return MPMDManager.__find_program_or_error(identifier).size

  @staticmethod
  def finalize():
    MPMDManager.__programs_data.clear()

    local_comm = MPMDManager.local_comm()
    if local_comm != MPI.COMM_NULL: local_comm.Disconnect()

    manager_comm = MPMDManager.__manager_comm
    if manager_comm != MPI.COMM_NULL: manager_comm.Disconnect()
  
  @staticmethod
  def __set_programs_data(local_name: str, gathered_names: list[str]):
    MPMDManager.__set_local_data(local_name, gathered_names)

    previous_name: str | None = None
    rank, length = 0, len(gathered_names)

    while rank < length:
      name = gathered_names[rank]
      rank += 1

      if name == previous_name: 
        MPMDManager.__programs_data[-1].increment_size()
      else:
        previous_name = name

        if name == local_name: 
          MPMDManager.__programs_data.append(MPMDManager.__local_data)
          rank += MPMDManager.local_size() - 1
        else: MPMDManager.__add_program(name, rank - 1)          

  @staticmethod
  def __set_local_data(local_name: str, gathered_names: list[str]):
    global_rank = MPMDManager.__manager_comm.Get_rank()
    local_id = gathered_names.index(local_name)

    local_comm = MPMDManager.__manager_comm.Split(local_id, global_rank)
    local_size = local_comm.Get_size()
    MPMDManager.__local_data = ProgramData(local_name, local_comm, local_size)

  @staticmethod
  def __add_program(name: str, initial_rank: int):
    intercomm = MPMDManager.__get_program_intercomm(initial_rank)
    MPMDManager.__programs_data.append(ProgramData(name, intercomm))

  
  @staticmethod
  def __get_program_intercomm(remote_program_initial_rank: int):
    return MPMDManager.local_comm().Create_intercomm(
      0, MPMDManager.__manager_comm, remote_program_initial_rank
    )
  
  @staticmethod
  def __find_program_or_error(identifier: str | int):
    invalid_identifier = not isinstance(identifier, (str, int))
    if invalid_identifier: raise Exception("Invalid program identifier!")

    program_found = MPMDManager.__find_program(identifier)
    if program_found is None: raise Exception("Program not found!")

    return program_found

  @staticmethod
  def __find_program(identifier: str | int):
    is_ind = isinstance(identifier, int)

    if is_ind: return MPMDManager.__find_program_by_ind(identifier)
    return MPMDManager.__find_program_by_name(identifier)
  
  @staticmethod
  def __find_program_by_ind(ind: int):
    invalid_ind = ind < 0 or ind >= len(MPMDManager.__programs_data)
    return None if invalid_ind else MPMDManager.__programs_data[ind]
  
  @staticmethod
  def __find_program_by_name(name: str):
    for data in MPMDManager.__programs_data: 
      if name == data.name: return data

File: src/single_world/c/abstraction/mpmdManager.c
#include "mpmdManager.h"

typedef struct {
  const char* name;
  uint leader, size;
} AuxiliarProgramData;

typedef struct {
  AuxiliarProgramData* programs;
  uint localProgramInd, quantity;
} AuxiliarData;

void setProgramsData(MPMDManager*, char**);

const char* gatherNames(MPI_Comm*, char**, uint);
void getAuxiliarData(AuxiliarData*, const char*, uint, uint);

void setLocalProgramData(MPMDManager*, AuxiliarData*, uint);
void setRemoteProgramsData(MPMDManager*, AuxiliarData*);

ProgramData* localProgram(const MPMDManager*);

ProgramData* findProgramOrError(
  const MPMDManager*, ProgramIdentifier, IdentifierType
);
ProgramData* findProgramByIndex(const MPMDManager*, uint);
ProgramData* findProgramByName(const MPMDManager*, const char*);

const MPMDManager* Manager_Init(char** argv) {
  MPMDManager* manager = malloc(sizeof(MPMDManager));
  MPI_Comm* managerComm = &manager->comm;
  MPI_Comm_dup(MPI_COMM_WORLD, managerComm);

  setProgramsData(manager, argv);
  return manager;
}

void Manager_Finalize(const MPMDManager* manager) {
  ProgramData* programs = manager->programsData;
  const uint programsQuantity = manager->programsQuantity;

  for(uint ind = 0; ind < programsQuantity; ind++) {
    MPI_Comm* programComm = &programs[ind].comm;
    if(programComm != NULL) MPI_Comm_disconnect(programComm);
  }
  free(manager->programsData);

  MPI_Comm* managerComm = (MPI_Comm*)&manager->comm;
  if(managerComm != NULL) MPI_Comm_disconnect(managerComm);

  free((MPMDManager*)manager);
}

const char* Manager_Local_Name(const MPMDManager* manager) {
  return localProgram(manager)->name;
}

uint Manager_Local_Size(const MPMDManager* manager) {
  return localProgram(manager)->size;
}

const MPI_Comm* Manager_Local_Comm(const MPMDManager* manager) {
  return &localProgram(manager)->comm;
}

const MPI_Comm* Manager_Intercomm_to(
  const MPMDManager* manager, ProgramIdentifier identifier,
  IdentifierType identifierType
) {
  return &findProgramOrError(manager, identifier, identifierType)->comm;
}

uint Manager_Size_of(
  const MPMDManager* manager, ProgramIdentifier identifier,
  IdentifierType identifierType
) {
  return findProgramOrError(manager, identifier, identifierType)->size;
}

void setProgramsData(MPMDManager* manager, char** argv) {
  AuxiliarData auxiliar;
  int worldRank, worldSize;
  MPI_Comm* managerComm = &manager->comm;

  MPI_Comm_rank(*managerComm, &worldRank);
  MPI_Comm_size(*managerComm, &worldSize);

  const char* gatheredNames = gatherNames(managerComm, argv, worldSize);
  getAuxiliarData(&auxiliar, gatheredNames, worldSize, worldRank);

  manager->programsData = malloc(sizeof(ProgramData) * auxiliar.quantity);
  manager->programsQuantity = auxiliar.quantity;

  setLocalProgramData(manager, &auxiliar, worldRank);
  setRemoteProgramsData(manager, &auxiliar);

  free(auxiliar.programs);
  free((char*)gatheredNames);
}

const char* gatherNames(MPI_Comm* managerComm, char** argv, uint worldSize) {
  int nameSize;
  char* gatheredNames = malloc(sizeof(char) * worldSize * NAME_MAX_SIZE);

  getFilename(gatheredNames, &nameSize, argv);
  MPI_Allgather(
    gatheredNames, nameSize, MPI_CHAR, gatheredNames, NAME_MAX_SIZE, MPI_CHAR,
    *managerComm
  );

  return gatheredNames;
}

void getAuxiliarData(
  AuxiliarData* auxiliarData, const char* gatheredNames, uint worldSize,
  uint worldRank
) {
  AuxiliarProgramData* data = malloc(sizeof(AuxiliarProgramData) * worldSize);
  auxiliarData->programs = data;

  int programInd = -1;
  const char* previousName = NULL;
  const char* currentName = gatheredNames;

  for(int ind = 0; ind < worldSize; ind++) {
    if(streql(previousName, currentName)) data[programInd].size++;
    else {
      data[++programInd] = (AuxiliarProgramData){currentName, ind, 1};
      if(ind <= worldRank) auxiliarData->localProgramInd = programInd;
    }

    previousName = currentName;
    currentName += NAME_MAX_SIZE;
  }
  auxiliarData->quantity = programInd + 1;
}

void setLocalProgramData(
  MPMDManager* manager, AuxiliarData* auxiliar, uint worldRank
) {
  MPI_Comm* managerComm = &manager->comm;
  uint localProgramInd = manager->localProgramInd = auxiliar->localProgramInd;

  ProgramData* localProgram = &manager->programsData[localProgramInd];
  AuxiliarProgramData* auxiliarLocal = &auxiliar->programs[localProgramInd];

  strcpy(localProgram->name, auxiliarLocal->name);
  localProgram->size = auxiliarLocal->size;
  MPI_Comm_split(*managerComm, localProgramInd, worldRank, &localProgram->comm);
}

void setRemoteProgramsData(MPMDManager* manager, AuxiliarData* auxiliar) {
  const uint programsQuantity = auxiliar->quantity;
  const uint localProgramInd = auxiliar->localProgramInd;
  ProgramData* localProgram = &manager->programsData[localProgramInd];

  for(uint ind = 0; ind < programsQuantity; ind++) {
    if(ind == localProgramInd) continue;

    ProgramData* currentProgram = &manager->programsData[ind];
    AuxiliarProgramData* currentAuxiliar = &auxiliar->programs[ind];

    strcpy(currentProgram->name, currentAuxiliar->name);
    currentProgram->size = currentAuxiliar->size;

    MPI_Intercomm_create(
      localProgram->comm, 0, manager->comm, currentAuxiliar->leader, 0,
      &currentProgram->comm
    );
  }
}

ProgramData* localProgram(const MPMDManager* manager) {
  return &manager->programsData[manager->localProgramInd];
}

ProgramData* findProgramOrError(
  const MPMDManager* manager, ProgramIdentifier identifier,
  IdentifierType identifierType
) {
  ProgramData* programFound;
  const bool indexId = identifierType == INDEX_ID;

  if(indexId) programFound = findProgramByIndex(manager, identifier.index);
  else programFound = findProgramByName(manager, identifier.name);

  if(programFound == NULL) MPI_Abort(manager->comm, EXIT_FAILURE);
  return programFound;
}

ProgramData* findProgramByIndex(const MPMDManager* manager, uint index) {
  if(index >= manager->programsQuantity) return NULL;
  return &manager->programsData[index];
}

ProgramData* findProgramByName(const MPMDManager* manager, const char* name) {
  ProgramData* programsData = manager->programsData;
  const uint programsQuantity = manager->programsQuantity;

  for(uint ind = 0; ind < programsQuantity; ind++) {
    ProgramData* currentProgram = &programsData[ind];
    if(streql(name, currentProgram->name)) return currentProgram;
  }

  return NULL;
}